[FEAT] support git worktrees

**Is your feature request related to a problem? Please describe.**

Using [git worktrees](https://git-scm.com/docs/git-worktree) (or more precisely, using multiple working trees attached to the same repository) becomes more and more common.

`husky` is currently not compatible with this setup. When running any `husky` program within a linked worktree, the program terminates with an error:

```
Installing hooks
panic: unlinkat .git/hooks: not a directory

goroutine 1 [running]:
github.com/automation-co/husky/cmd.init.func2(0x73abf5a1a900?, {0x100359f2f?, 0x4?, 0x100359f33?})
	/Users/bastian/go/pkg/mod/github.com/automation-co/husky@v0.2.16/cmd/init.go:19 +0x34
github.com/spf13/cobra.(*Command).execute(0x100551ce0, {0x10057c8a0, 0x0, 0x0})
	/Users/bastian/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:860 +0x55c
github.com/spf13/cobra.(*Command).ExecuteC(0x1005521e0)
	/Users/bastian/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:974 +0x2e4
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/bastian/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:902
github.com/automation-co/husky/cmd.Execute()
	/Users/bastian/go/pkg/mod/github.com/automation-co/husky@v0.2.16/cmd/root.go:20 +0x24
main.main()
	/Users/bastian/go/pkg/mod/github.com/automation-co/husky@v0.2.16/main.go:6 +0x1c
```

This is because in a linked worktree, `.git` is actually a file, not a directory. Husky only checks whether `.git` exists, then blindly assumes that it is a directory, and creates the path to `.git/hooks` by concatenating `.git` + `/hooks`.

**Describe the solution you'd like**
Support multi-worktree setups in addition to single-worktree setups.

**Describe alternatives you've considered**
-

**Additional context**
I've already forked the repo and basically implemented support for worktrees, I'll put a PR up soon.

--- a/README.md
+++ b/README.md
@@ -35,6 +35,11 @@ go install github.com/automation-co/husky@latest
 
 ### Getting Started
 
+*Warning:* All `husky` commands are *potentially destructive*, they will *delete* all files in
+`.git/hooks` and recreate them from `.husky/hooks`.
+If you already have custom hooks in `.git/hooks`, make sure to create a backup of the hooks before
+starting to use `husky`.
+
 You can initialise husky by `$ husky init`
 
 > Make sure you have git initialised

--- a/cmd/add.go
+++ b/cmd/add.go
@@ -20,11 +20,13 @@ echo 'woof'
 		cmdStr := args[1]
 
 		if err := lib.Add(hook, cmdStr); err != nil {
-			panic(err)
+			exitOnError(err)
+			return
 		}
 
 		if err := lib.Install(); err != nil {
-			panic(err)
+			exitOnError(err)
+			return
 		}
 	},
 }

--- a/cmd/init.go
+++ b/cmd/init.go
@@ -16,7 +16,8 @@ https://github.com/automation-co/husky
 `,
 	Run: func(cmd *cobra.Command, args []string) {
 		if err := lib.Init(); err != nil {
-			panic(err)
+			exitOnError(err)
+			return
 		}
 	},
 }

--- a/cmd/install.go
+++ b/cmd/install.go
@@ -11,7 +11,8 @@ var installCmd = &cobra.Command{
 	Long:  `Install hooks from the .hooks folder`,
 	Run: func(cmd *cobra.Command, args []string) {
 		if err := lib.Install(); err != nil {
-			panic(err)
+			exitOnError(err)
+			return
 		}
 	},
 }

--- a/cmd/util.go
+++ b/cmd/util.go
@@ -1,0 +1,11 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+)
+
+func exitOnError(err error) {
+	fmt.Printf("error: %v\n", err)
+	os.Exit(1)
+}

--- a/internal/lib/add.go
+++ b/internal/lib/add.go
@@ -1,61 +1,54 @@
 package lib
 
 import (
-	"errors"
 	"fmt"
 	"os"
-	"path"
+	"path/filepath"
+	"strings"
 )
 
-// Add command will create the file from hook value into .husky/hooks directory. The cmd appended to shebang string and
-// written in the file .husky/hooks/<hook>. The function intend to fail if the git hook name is invalid.
+// Add creates a hook file in .husky/hooks under the current worktree.
+// The cmd argument is prefixed with a #!/bin/sh shebang and written to
+// .husky/hooks/<hook>. Fails if the hook name is invalid, the current
+// directory is not inside a git working tree, or .husky is not initialized.
 func Add(hook string, cmd string) error {
-	// validate hooks
 	if !isValidHook(hook) {
-		return errors.New("invalid hook name")
+		return fmt.Errorf("invalid hook name %s, valid hooks are: %v", hook, strings.Join(validHooks, ", "))
 	}
 
-	// check if .git exists
-	if isExists, err := gitExists(); err == nil && !isExists {
-		return errors.New("git not initialized")
-	} else if err != nil {
-		return err
-	}
-
-	// check if .husky exists
-	if isExists, err := huskyExists(); err == nil && !isExists {
-		return errors.New(".husky not initialized")
-	} else if err != nil {
-		return err
-	}
-
-	// check if .husky/hooks exists
-	_, err := os.Stat(getHuskyHooksDir(true))
-	fmt.Println(err)
-	if os.IsNotExist(err) {
-		fmt.Println("no pre-existing hooks found")
-
-		// create .husky/hooks
-		err = os.MkdirAll(getHuskyHooksDir(true), 0755)
-		if err != nil {
-			return err
-		}
-
-		fmt.Println("created .husky/hooks")
-	}
-
-	// create hook
-	file, err := os.Create(path.Join(getHuskyHooksDir(true), hook))
+	// check if .git exists and find the root directory of the current git worktree
+	worktreeRoot, err := renderWorktreeRoot()
 	if err != nil {
 		return err
 	}
 
+	// check if the .husky directory exists
+	huskyDir := renderHuskyDir(worktreeRoot)
+	if _, err := os.Stat(huskyDir); os.IsNotExist(err) {
+		return fmt.Errorf("the directory %s does not exist, run \"husky init\" and try again", huskyDir)
+	} else if err != nil {
+		return err
+	}
+
+	huskyHooksDir := renderHuskyHooksDir(worktreeRoot)
+	if _, err := os.Stat(huskyHooksDir); os.IsNotExist(err) {
+		fmt.Printf("no pre-existing hooks found, creating %s now\n", huskyHooksDir)
+		if err := os.MkdirAll(huskyHooksDir, 0755); err != nil {
+			return err
+		}
+		fmt.Printf("created %s\n", huskyHooksDir)
+	} else if err != nil {
+		return err
+	}
+
+	file, err := os.Create(filepath.Join(huskyHooksDir, hook))
+	if err != nil {
+		return err
+	}
 	//goland:noinspection GoUnhandledErrorResult
 	defer file.Close()
 
-	cmd = "#!/bin/sh\n" + cmd
-	_, err = file.WriteString(cmd)
-	if err != nil {
+	if _, err := file.WriteString("#!/bin/sh\n" + cmd); err != nil {
 		return err
 	}
 

--- a/internal/lib/init.go
+++ b/internal/lib/init.go
@@ -1,53 +1,46 @@
 package lib
 
 import (
-	"errors"
+	"fmt"
 	"os"
-	"path"
+	"path/filepath"
 )
 
-// Init command will set up the .husky directory as sibling of .git directory if not exists install pre-commit hook by default
-// If .husky exists, it will remove all the files from .git/hooks directory and copy from .husky directory.
+// Init sets up the .husky/hooks directory under the current worktree if it
+// does not already exist, creates a default pre-commit hook, and then calls
+// Install() to link the hooks into git's hooks directory.
 func Init() error {
-	// check if .git exists
-	if isExists, err := gitExists(); err == nil && !isExists {
-		return errors.New("git not initialized")
-	} else if err != nil {
+	worktreeRoot, err := renderWorktreeRoot()
+	if err != nil {
 		return err
 	}
 
 	// check if .husky exists
-	if isExists, err := huskyExists(); err == nil && isExists {
-		return errors.New(".husky already exist")
-	} else if err != nil {
-		return err
+	huskyDir := renderHuskyDir(worktreeRoot)
+	if _, err = os.Stat(huskyDir); err == nil {
+		return fmt.Errorf("%s already exists", huskyDir)
+	} else if !os.IsNotExist(err) {
+		return fmt.Errorf("error when checking if %s already exists: %v", huskyDir, err)
 	}
 
-	// if not, create .husky/hooks
-	err := os.MkdirAll(getHuskyHooksDir(true), 0755)
+	// create .husky/hooks
+	huskyHooks := renderHuskyHooksDir(worktreeRoot)
+	fmt.Printf("creating the %s now\n", huskyHooks)
+	if err := os.MkdirAll(huskyHooks, 0755); err != nil {
+		return fmt.Errorf("error when creating %s: %v", huskyHooks, err)
+	}
+
+	preCommitHook := filepath.Join(huskyHooks, "pre-commit")
+	file, err := os.Create(preCommitHook)
 	if err != nil {
-		return err
+		return fmt.Errorf("error when creating %s: %v", preCommitHook, err)
 	}
-
-	// create default pre-commit hook
-	file, err := os.Create(path.Join(getHuskyHooksDir(true), "pre-commit"))
-	if err != nil {
-		return err
-	}
-
 	//goland:noinspection GoUnhandledErrorResult
 	defer file.Close()
 
-	_, err = file.WriteString(`#!/bin/sh`)
-	if err != nil {
-		return err
+	if _, err := file.WriteString("#!/bin/sh\n"); err != nil {
+		return fmt.Errorf("error when writing to %s: %v", preCommitHook, err)
 	}
 
-	// add hooks to .git/hooks
-	err = Install()
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return Install()
 }

--- a/internal/lib/install.go
+++ b/internal/lib/install.go
@@ -7,75 +7,64 @@ import (
 	"path/filepath"
 )
 
-// Install command will iterate the files from .husky/hooks create the hardlinks into .git/hooks and chmod to 0755 permission.
-// The function intend to fail in case .git or .husky directory doesn't exist.
+// Install iterates over the files in .husky/hooks, creates hardlinks into git's
+// hooks directory, and chmods them to 0755. Fails if the current directory
+// is not inside a git working tree or if .husky is not initialized.
 func Install() error {
 	fmt.Println("Installing hooks")
 
-	// check if .git exists
-	if isExists, err := gitExists(); err == nil && !isExists {
-		return errors.New("git not initialized")
-	} else if err != nil {
-		return err
-	}
-
-	// check if .husky exists
-	if isExists, err := huskyExists(); err == nil && !isExists {
-		return errors.New(".husky not initialized")
-	} else if err != nil {
-		return err
-	}
-
-	gitHooksDir, huskyHooksDir := getGitHooksDir(true), getHuskyHooksDir(true)
-	// check if .husky/hooks exists
-	_, err := os.Stat(huskyHooksDir)
-	if os.IsNotExist(err) {
-		return errors.New("no hooks found")
-	}
-
-	// delete all files in .git/hooks
-	if err := os.RemoveAll(gitHooksDir); err != nil {
-		return err
-	}
-
-	// create .git/hooks
-	if err := os.Mkdir(gitHooksDir, 0755); err != nil {
-		return err
-	}
-
-	// copy all files in .husky/hooks to .git/hooks
-	var hooks []string
-	err = filepath.Walk(huskyHooksDir,
-		func(path string, info os.FileInfo, err error) error {
-			hooks = append(hooks, path)
-			return nil
-		})
+	worktreeRoot, err := renderWorktreeRoot()
 	if err != nil {
 		return err
 	}
-	for _, hook := range hooks {
 
-		// skip .husky/hooks
-		if hook == huskyHooksDir {
+	huskyDir := renderHuskyDir(worktreeRoot)
+	if _, err := os.Stat(huskyDir); os.IsNotExist(err) {
+		return errors.New(".husky not initialized")
+	} else if err != nil {
+		return fmt.Errorf("error when checking if %s already exists: %v", huskyDir, err)
+	}
+
+	huskyHooks := renderHuskyHooksDir(worktreeRoot)
+	if _, err := os.Stat(huskyHooks); os.IsNotExist(err) {
+		return fmt.Errorf("no hooks found in %s, use \"husky add\" to create some hooks first", huskyHooks)
+	} else if err != nil {
+		return fmt.Errorf("error when checking if %s already exists: %v", huskyHooks, err)
+	}
+
+	gitHooks, cmd, err := renderGitHooksDir(worktreeRoot)
+	if err != nil {
+		return fmt.Errorf("error when executing \"%v\" to find .git/hooks: %v", cmd, err)
+	}
+
+	fmt.Printf("deleting %s now\n", gitHooks)
+	if err := os.RemoveAll(gitHooks); err != nil {
+		return err
+	}
+	fmt.Printf("re-creating %s\n", gitHooks)
+	if err := os.MkdirAll(gitHooks, 0755); err != nil {
+		return err
+	}
+
+	entries, err := os.ReadDir(huskyHooks)
+	if err != nil {
+		return fmt.Errorf("error when listing the directory %s: %v", huskyHooks, err)
+	}
+	for _, entry := range entries {
+		if entry.IsDir() {
 			continue
 		}
-
-		fmt.Println(hook)
-
-		// copy file to .git/hooks
-		err = os.Link(hook, filepath.Join(gitHooksDir, filepath.Base(hook)))
-		if err != nil {
-			return err
+		src := filepath.Join(huskyHooks, entry.Name())
+		dst := filepath.Join(gitHooks, entry.Name())
+		if err := os.Link(src, dst); err != nil {
+			return fmt.Errorf("error when hard-linking from %s to %s: %v", src, dst, err)
 		}
-
-		// make file executable
-		err = os.Chmod(filepath.Join(gitHooksDir, filepath.Base(hook)), 0755)
-		if err != nil {
-			return err
+		if err := os.Chmod(dst, 0755); err != nil {
+			return fmt.Errorf("error making %s executable: %v", dst, err)
 		}
-
+		fmt.Printf("hard-link for %s -> %s has been created\n", src, dst)
 	}
-	fmt.Println("Hooks installed")
 
+	fmt.Println("All hooks installed successfully")
 	return nil
 }

--- a/internal/lib/lib_test.go
+++ b/internal/lib/lib_test.go
@@ -1,7 +1,6 @@
 package lib_test
 
 import (
-	huskyLib "github.com/automation-co/husky/internal/lib"
 	"io/ioutil"
 	"math/rand"
 	"os"
@@ -10,11 +9,13 @@ import (
 	"strings"
 	"testing"
 	"time"
+
+	huskyLib "github.com/automation-co/husky/internal/lib"
 )
 
 // global error message declration block
 const (
-	nilErrorMsg             = "method has return nil error"
+	nilErrorMsg             = "method has returned nil error"
 	invalidContentsErrorMsg = "invalid file contents"
 	expectedDirErrorMsg     = "expected the path to be a directory"
 	expectedFileErrorMsg    = "expected the path to be a file"
@@ -86,7 +87,7 @@ func TestAddNoGit(t *testing.T) {
 	err = huskyLib.Add("pre-commit", "whoami")
 	if err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != "git not initialized" {
+	} else if !strings.HasPrefix(err.Error(), "this does not seem to be a git repository") {
 		t.Error(err)
 	}
 
@@ -116,7 +117,7 @@ func TestAddNoHusky(t *testing.T) {
 
 	if err := huskyLib.Add("pre-commit", "whoami"); err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != ".husky not initialized" {
+	} else if !strings.Contains(err.Error(), ".husky does not exist, run \"husky init\" and try again") {
 		t.Error(err)
 	}
 }
@@ -173,7 +174,7 @@ func TestInitNoGit(t *testing.T) {
 	err = huskyLib.Init()
 	if err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != "git not initialized" {
+	} else if !strings.HasPrefix(err.Error(), "this does not seem to be a git repository") {
 		t.Error(err)
 	}
 
@@ -207,7 +208,7 @@ func TestInitHuskyExists(t *testing.T) {
 
 	if err := huskyLib.Init(); err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != ".husky already exist" {
+	} else if !strings.Contains(err.Error(), ".husky already exist") {
 		t.Error(err)
 	}
 }
@@ -268,7 +269,7 @@ func TestInstallNoGit(t *testing.T) {
 	err = huskyLib.Install()
 	if err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != "git not initialized" {
+	} else if !strings.HasPrefix(err.Error(), "this does not seem to be a git repository") {
 		t.Error(err)
 	}
 
@@ -325,7 +326,7 @@ func TestInstallNoHuskyHooks(t *testing.T) {
 
 	if err := huskyLib.Install(); err == nil {
 		t.Error(nilErrorMsg)
-	} else if err.Error() != "no hooks found" {
+	} else if !strings.Contains(err.Error(), "no hooks found in ") {
 		t.Error(err)
 	}
 }

--- a/internal/lib/utils.go
+++ b/internal/lib/utils.go
@@ -1,8 +1,10 @@
 package lib
 
 import (
-	"os"
-	"path"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
 )
 
 // git hooks currently supported
@@ -39,50 +41,44 @@ func isValidHook(hook string) bool {
 	return contains(validHooks, hook)
 }
 
-// gitExists will return true if the comman is executed under .git directory
-// TODO: support recursive find .git directory till home
-func gitExists() (bool, error) {
-	// check if .git exists
-	_, err := os.Stat(".git")
-	if os.IsNotExist(err) {
-		return false, nil
-	} else if err != nil {
-		return false, err
+// renderWorktreeRoot returns the absolute path of the top-level directory of the
+// current git working tree. In plain vanilla repository clones with only one
+// worktree, this is the repository root directory. In clones with multiple linked
+// worktrees, this is the linked worktree's own root, not the main repository's working
+// tree. renderWorktreeRoot assumes to be called from inside the worktree.
+func renderWorktreeRoot() (string, error) {
+	out, err := exec.Command("git", "rev-parse", "--show-toplevel").Output()
+	if err != nil {
+		return "", fmt.Errorf("this does not seem to be a git repository, make sure to call this from within the repository - or from within a worktree if you are using worktrees; error: %v", err)
 	}
-
-	return true, nil
+	return strings.TrimSpace(string(out)), nil
 }
 
-// huskyExists will return true if exists, otherwise false
-// TODO: support recursive find .husky directory till home
-func huskyExists() (bool, error) {
-	_, err := os.Stat(".husky")
-
-	if os.IsNotExist(err) {
-		return false, nil
-	} else if err != nil {
-		return false, err
+// renderGitHooksDir returns the absolute path to the hooks directory git will
+// actually execute from. Honors linked worktrees (→ shared common dir),
+// submodules, and core.hooksPath, because `git rev-parse --git-path hooks`
+// does.
+func renderGitHooksDir(worktreeRoot string) (string, exec.Cmd, error) {
+	cmd := exec.Command("git", "rev-parse", "--git-path", "hooks")
+	cmd.Dir = worktreeRoot
+	out, err := cmd.Output()
+	if err != nil {
+		return "", *cmd, err
 	}
-
-	return true, nil
+	p := strings.TrimSpace(string(out))
+	if !filepath.IsAbs(p) {
+		p = filepath.Join(worktreeRoot, p)
+	}
+	return p, *cmd, nil
 }
 
-// getHuskyHooksDir will return the relative or absolute .husky hooks directory
-func getHuskyHooksDir(relative bool) string {
-	if relative {
-		return path.Join(".husky", "hooks")
-	}
-
-	cwd, _ := os.Getwd()
-	return path.Join(cwd, ".husky", "hooks")
+// renderHuskyDir returns the absolute path to the .husky directory inside the
+// worktree root directory worktreeRoot.
+func renderHuskyDir(worktreeRoot string) string {
+	return filepath.Join(worktreeRoot, ".husky")
 }
 
-// getGitHooksDir will return the relative or absolute .git hooks directory
-func getGitHooksDir(relative bool) string {
-	if relative {
-		return path.Join(".git", "hooks")
-	}
-
-	cwd, _ := os.Getwd()
-	return path.Join(cwd, ".git", "hooks")
+// renderHuskyHooksDir returns the absolute path to .husky/hooks inside worktreeRoot.
+func renderHuskyHooksDir(worktreeRoot string) string {
+	return filepath.Join(renderHuskyDir(worktreeRoot), "hooks")
 }

